### PR TITLE
Manage stops

### LINKED ISSUE #18: Manage stops
Would be useful to know when stops have been updated. Some common uses:

* How many stops have not been updated since X date
* A list of the stops that have not been updated since X date

Might be more flexible to provide:

* How many stops were updated over a date range
* A list of stops that were updated over a date range

Or a CSV export

### PR DESCRIPTION
**Pull request by dfaulken**
Closes #18.

Link to new interface in navbar:

screen shot 2018-03-20 at 5 41 58 pm

Management interface:

screen shot 2018-03-20 at 5 42 16 pm

Includes pagination:

screen shot 2018-03-20 at 5 42 31 pm

Can view outdated stops:

screen shot 2018-03-20 at 5 42 48 pm

Including a datepicker:

screen shot 2018-03-20 at 5 43 18 pm

And can export as CSV:

screen shot 2018-03-20 at 5 43 43 pm

### CODE REVIEW

**Comment by dfaulken** [comment]
Can update all as generic CSV:

screen shot 2018-03-20 at 5 53 54 pm

Link:

screen shot 2018-03-20 at 5 54 10 pm

File: Gemfile [hunk 1]
```diff
@@ -23,4 +24,5 @@ group :development do
   gem 'capistrano-rails', require: false
   gem 'capistrano-passenger', require: false
   gem 'pry-byebug'
+  gem 'timecop', '~> 0.9.1'
```
Comment (dfaulken) [inline Gemfile]: Timecop is used in seeding to simulate incidents having been created/updated at different times in the past.

File: app/assets/javascripts/bus_stops.js [hunk 2]
```diff
@@ -2,6 +2,10 @@ $(document).ready(function(){
   $('.name-autocomplete').autocomplete({
     source: searchStops
   });
+
+  $('.datepick').datepicker({
+    dateFormat: 'yy-mm-dd'
```
Comment (dfaulken) [inline app/assets/javascripts/bus_stops.js]: It *looks* like a two-digit year, but it's actually four. If you write `yyyy` you end up with `20182018`.

File: app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- *= require autocomplete
+ *= require jquery-ui
```
Comment (dfaulken) [inline app/assets/stylesheets/application.scss]: Just all of it, why not

File: app/controllers/bus_stops_controller.rb [hunk 4]
```diff
@@ -1,16 +1,17 @@
 class BusStopsController < ApplicationController
-  before_action :find_stop, only: %i(edit id_search update)
+  before_action :find_stop, only: %i[edit destroy id_search update]
+  before_action :restrict_to_admin, only: %i[destroy manage]
 
   def autocomplete
     stops = BusStop.where 'lower(name) like ?',
                           "%#{params.require(:term)}%"
-    render json: stops.pluck(:name)
+    render json: stops.pluck(:name).sort
```
Comment (dfaulken) [inline app/controllers/bus_stops_controller.rb]: Due to removing default scopes (I don't like them anymore)

File: app/models/bus_stop.rb [hunk 5]
```diff
@@ -50,6 +60,23 @@ def last_updated_by
...
   end
 
+  def route_list
+    routes.pluck(:number).sort.join(', ')
+  end
+
+  def self.to_csv(limited_attributes: false)
+    attrs = if limited_attributes
+              LIMITED_ATTRS
+            else Hash[columns.map{|c| [c.name, c.name.humanize] }]
```
Comment (dfaulken) [inline app/models/bus_stop.rb]: Should I use `human_attribute_name` here to respect locales in the future?

File: app/models/bus_stop.rb [hunk 6]
```diff
@@ -50,6 +60,23 @@ def last_updated_by
...
+            end
+    CSV.generate headers: true do |csv|
+       csv << attrs.values
+       all.each do |stop|
+         csv << attrs.keys.map { |attr| stop.send attr }
+       end
```
Comment (dfaulken) [inline app/models/bus_stop.rb]: I wanted to do

```ruby
csv += all.map do |stop|
  attrs.keys.map { |attr| stop.send.attr }
end
```,

but `CSV` doesn't implement `+=`.

File: app/views/bus_stops/outdated.haml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+Bus stops not updated since #{@date} (#{@stops.count})
+
+= form_tag '#', method: :get do
+  = text_field_tag :date, @date, class: 'datepick'
+  = submit_tag 'Change date'
+
+= link_to 'Export as CSV', outdated_bus_stops_path(date: params[:date], format: :csv)
```
Comment (dfaulken) [inline app/views/bus_stops/outdated.haml]: I tried `..._path(params, format: :csv)` for future proofing sake, but that didn't work.

File: app/models/bus_stop.rb (hunk 5 above)
Comment (Anbranin) [reply app/models/bus_stop.rb]: I wouldn't worry about it.

**Comment by dfaulken** [comment]
@sherson does the interface look good to you?

File: app/assets/javascripts/bus_stops.js (hunk 2 above)
Comment (mboneil10) [reply app/assets/javascripts/bus_stops.js]: Waaaaaat

File: app/controllers/bus_stops_controller.rb [hunk 8]
```diff
@@ -40,6 +59,20 @@ def name_search
     end
   end
 
+  def outdated
+    @date = Date.parse(params[:date]) rescue 1.month.ago.to_date
```
Comment (mboneil10) [inline app/controllers/bus_stops_controller.rb]: Why is the rescued value `1.month.ago.to_date`?

File: app/models/bus_stop.rb (hunk 6 above)
Comment (mboneil10) [reply app/models/bus_stop.rb]: :(

File: app/controllers/bus_stops_controller.rb (hunk 8 above)
Comment (dfaulken) [reply app/controllers/bus_stops_controller.rb]: It needs to be something. I'm open to suggestions.

**Comment by dfaulken** [comment]
> Do we need to use console magic to restore a mistakenly deleted stop?

Yes. Stops are paper-trailed, so that would be easiest.

**Pull request merged**